Cache mailer view

st0012 opened the pull request:
Hi, I'm adding cache feature to `ActionMailer`'s templates. I think there two ways to accomplish this goal. 
First is moving all the caching feature from `ActionController` to `AbstractController` since both `ActionMailer` and `ActionController` are inherited from it.
The second one is porting caching features to `ActionMailer`, which I think will break the `DRY` rule.
However, I'm working on this with the second approach. Because I think after I reimplemented it in `ActionMailer`, I can find some reusable parts in both `ActionMailer` and `ActionController` and put them into somewhere like `ActiveSupport`?
And I know this could be a huge work while I am not every familiar with Rails' codebase. So I am hoping someone can discuss this with me or give me some advice. Any suggestion will be very helpful, thanks!

st0012 committed:
Setting up test

st0012 committed:
Basic test setup done

st0012 committed:
Use ActionMailer::Caching to contain caching related methods

st0012 committed:
Porting ActionController::Caching to ActionMailer::Caching

st0012 committed:
Port some basic fragment tests

st0012 committed:
Fixing CacheHelperOutputBufferTest, ViewCacheDependencyTest and AutomaticCollectionCacheTest

st0012 committed:
Remove useless tests and fix ViewCacheDependencyTest

tadeusrox commented on actionmailer/.byebug_history:
```diff
@@ -0,0 +1,8 @@
+exit
```
I think this file shouldn't be here

If you think it's usefull -> https://help.github.com/articles/ignoring-files/#create-a-global-gitignore

st0012 replied:
@tadeusrox  Thanks!

st0012 committed:
Delete .byebug_history

st0012 commented:
And I think the main difference between `ActionController` and `ActionMailer`'s caching implementation is how the cache key is generated and `ActionMailer` doesn't need to deal with requests.

rafaelfranca commented:
> I can find some reusable parts in both ActionMailer and ActionController and put them into somewhere like ActiveSupport?

Yes. You can put in Action Dispatch like Abstract Controller already is.

st0012 committed:
Adding functional test

st0012 committed:
Update test template

st0012 committed:
Passed first functional test

st0012 committed:
Use caching mailer to do caching tests

st0012 committed:
Improve functional test and add partial test

st0012 committed:
Add skip_fragment_cache_digesting test

st0012 commented:
Hi, I'm wondering if I need to copy all the caching test from `ActionController`, and change them to the `ActionMailer` version? I think most of them are testing the fragment cache itself, not testing `fragment cache in ActionController`. So make a copy of them in `ActionMailer` are basically unnecessary. What do you think? cc @rafaelfranca @arthurnn

st0012 committed:
Remove unused tests

st0012 committed:
Move caching/fragments in ActionMailer and ActionController to action_dispatch/caching/fragments

st0012 committed:
Move most caching methods to ActionDispatch::Caching, and let ActionMailer and ActionController to include it

st0012 committed:
Fix CI

st0012 commented:
I don't know why CI would fail. It passed in my local VM with command: `GEM=am,amo,as,av,aj bundle exec rake test`.

st0012 committed:
Make caching configuration more flexible

st0012 commented:
@arthurnn Hi, can you please give me some advice? thank you :smile:

rafaelfranca commented:
> Hi, I'm wondering if I need to copy all the caching test from ActionController, and change them to the ActionMailer version? I think most of them are testing the fragment cache itself, not testing fragment cache in ActionController. So make a copy of them in ActionMailer are basically unnecessary. What do you think?

Yes, agree. It is better to have a fresh start in the test suite, but it is good to use the ActionController tests as inspiration.

rafaelfranca commented on actionmailer/lib/action_mailer/caching.rb:
```diff
@@ -0,0 +1,22 @@
+
+module ActionMailer
+  module Caching
+    extend ActiveSupport::Autoload
```
you don't need this.

rafaelfranca commented on actionmailer/lib/action_mailer/caching.rb:
```diff
@@ -0,0 +1,22 @@
+require 'action_dispatch/caching'
+
+module ActionMailer
+  module Caching
```
I think we can remove this module and just include its content directly in `ActionMailer::Base`

st0012 replied:
Should we also move `ActionController::Caching` into `ActionController::Base`?

rafaelfranca commented on actionpack/lib/action_controller/caching.rb:
```diff
@@ -23,65 +22,24 @@ module ActionController
   #   config.action_controller.cache_store = MyOwnStore.new('parameter')
   module Caching
-    extend ActiveSupport::Concern
     extend ActiveSupport::Autoload
```
This line can be removed.

st0012 replied:
Remove this line will cause `undefined method eager_load!` error in [here](https://github.com/rails/rails/blob/master/actionpack%2Flib%2Faction_controller.rb#L49).

rafaelfranca commented on actionpack/lib/action_dispatch/caching.rb:
```diff
@@ -0,0 +1,61 @@
+
+    protected
+      # Convenience accessor.
+    def cache(key, options = {}, &block)
```
The indentation here needs to be fixed

rafaelfranca commented on actionpack/lib/action_dispatch/caching.rb:
```diff
@@ -0,0 +1,61 @@
+        end
+    end
+
+    include AbstractController::Helpers
```
Why is this included here? It is being used here somewhere? I think it is fine to remove since both classes that include `AD::Caching` already include this module.

st0012 committed:
Move ActionMailer::Caching's content into ActionMailer::Base instead of including it

Remove useless helper in ActionDispatch::Caching and fix indentation

st0012 committed:
Remove unnecessarily included modules in ActionController::Caching

st0012 committed:
Change ActionMailer's default caching configuration and update generator's environment templates

st0012 commented:
Here's what I have done for now:
1. I move the `ActionController::Caching` to `ActionDispatch::Caching` and include it in `ActionMailer::Base` and `ActionController::Caching`.
2. I added some unit tests and functional test to make sure that `ActionMailer` has the caching ability.
3. Finally, I changed the environment template files to add `ActionMailer`'s caching configuration.

And I am wondering what I should do next? This is my first time contributing `Rails`, so I think I need some guide to finish this :cry: .

st0012 commented:
@rafaelfranca Hello ~ what can I do to merge this?

st0012 commented:
@rafaelfranca Thanks for merging!

st0012 commented:
@rafaelfranca BTW, when should/can I edit the documentations?

rafaelfranca commented:
You can open a PR upgrading the documentation, and we really should.

st0012 commented:
@rafaelfranca ok, thanks!
